diff -up libibverbs-1.1.8/src/sysfs.c.coverity libibverbs-1.1.8/src/sysfs.c
--- libibverbs-1.1.8/src/sysfs.c.coverity	2014-10-09 15:50:17.611302631 -0400
+++ libibverbs-1.1.8/src/sysfs.c	2014-10-09 15:54:47.045695663 -0400
@@ -96,8 +96,19 @@ int ibv_read_sysfs_file(const char *dir,
 	close(fd);
 	free(path);
 
-	if (len > 0 && buf[len - 1] == '\n')
-		buf[--len] = '\0';
+	if (len > 0) {
+		if (buf[len - 1] == '\n')
+			buf[--len] = '\0';
+		else if (len < size)
+			buf[len] = '\0';
+		else
+			/* We would have to truncate the contents to NULL
+			 * terminate, so we are going to fail no matter
+			 * what we do, either right now or later when
+			 * we pass around an unterminated string.  Fail now.
+			 */
+			return -1;
+	}
 
 	return len;
 }
