diff -up libibverbs-1.1.6/examples/devinfo.c.error libibverbs-1.1.6/examples/devinfo.c
--- libibverbs-1.1.6/examples/devinfo.c.error	2011-12-21 20:11:09.000000000 -0500
+++ libibverbs-1.1.6/examples/devinfo.c	2012-04-05 15:36:08.512891199 -0400
@@ -218,10 +218,16 @@ static int print_hca_cap(struct ibv_devi
 		goto cleanup;
 	}
 	if (ibv_query_device(ctx, &device_attr)) {
-		fprintf(stderr, "Failed to query device props");
+		fprintf(stderr, "Failed to query device props\n");
 		rc = 2;
 		goto cleanup;
 	}
+	if (ib_port && ib_port > device_attr.phys_port_cnt) {
+		fprintf(stderr, "Invalid port requested for device\n");
+		/* rc = 3 is taken by failure to clean up */
+		rc = 4;
+		goto cleanup;
+	}
 
 	printf("hca_id:\t%s\n", ibv_get_device_name(ib_dev));
 	printf("\ttransport:\t\t\t%s (%d)\n",
@@ -379,7 +385,7 @@ int main(int argc, char *argv[])
 
 		case 'i':
 			ib_port = strtol(optarg, NULL, 0);
-			if (ib_port < 0) {
+			if (ib_port <= 0) {
 				usage(argv[0]);
 				return 1;
 			}
